File: src/main/java/fr/yodamad/svn2git/service/HistoryManager.java
package fr.yodamad.svn2git.service;

import fr.yodamad.svn2git.domain.Migration;
import fr.yodamad.svn2git.domain.MigrationHistory;
import fr.yodamad.svn2git.domain.enumeration.StatusEnum;
import fr.yodamad.svn2git.domain.enumeration.StepEnum;
import fr.yodamad.svn2git.repository.MigrationHistoryRepository;
import fr.yodamad.svn2git.repository.MigrationRepository;
import fr.yodamad.svn2git.service.util.DateFormatter;
import org.hibernate.Hibernate;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;

import static java.lang.String.format;

/**
 * Migration history operations
 */
@Service
public class HistoryManager {

    private static final Logger LOG = LoggerFactory.getLogger(HistoryManager.class);
    private final MigrationHistoryRepository migrationHistoryRepository;
    private final MigrationRepository migrationRepository;

    public HistoryManager(final MigrationHistoryRepository mhr, final MigrationRepository mr) {
        this.migrationHistoryRepository = mhr;
        this.migrationRepository = mr;
    }

    /**
     * Create a new history for migration
     * @param migration
     * @param step
     * @param data
     * @return
     */
    public MigrationHistory startStep(Migration migration, StepEnum step, String data) {
        MigrationHistory history = new MigrationHistory()
            .step(step)
            .migration(migration)
            .date(Instant.now())
            .status(StatusEnum.RUNNING)
            .startTime(Instant.now());

        if (data != null) {
            history.data(data);
        }

        LOG.info(format("Start step %s", step));
        return migrationHistoryRepository.save(history);
    }

    /**
     * Update history
     * @param history
     */
    public void endStep(MigrationHistory history, StatusEnum status, String data) {
        history.setStatus(status);
        if (data != null) history.setData(data);

        if (history.getStartTime() == null) { history.setExecutionTime("N/A"); }
        else {
            // Compute executionTime
            Long execution = Instant.now().toEpochMilli() - history.getStartTime().toEpochMilli();
            history.setExecutionTime(DateFormatter.toNiceFormat(execution));
        }

        migrationHistoryRepository.save(history);
        LOG.info(format("Finish step %s with status %s in %s", history.getStep(), status, history.getExecutionTime()));
    }

    /**
     * Load eagerly a migration
     * @param migId Migration ID
     * @return migration loaded
     */
    @Transactional
    public Migration loadMigration(Long migId) {
        Migration migration = migrationRepository.getOne(migId);
        Hibernate.initialize(migration);
        Hibernate.initialize(migration.getHistories());
        Hibernate.initialize(migration.getMappings());
        return migration;
    }

    /**
     * Force flush for summary
     */
    public void forceFlush() {
        migrationHistoryRepository.flush();
        migrationRepository.flush();
    }
}

File: src/main/kotlin/fr/yodamad/svn2git/service/HistoryManager.kt
package fr.yodamad.svn2git.service

import fr.yodamad.svn2git.domain.Migration
import fr.yodamad.svn2git.domain.MigrationHistory
import fr.yodamad.svn2git.domain.enumeration.StatusEnum
import fr.yodamad.svn2git.domain.enumeration.StepEnum
import fr.yodamad.svn2git.repository.MigrationHistoryRepository
import fr.yodamad.svn2git.repository.MigrationRepository
import fr.yodamad.svn2git.service.util.DateFormatter
import org.hibernate.Hibernate
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant

/**
 * Migration history operations
 */
@Service
open class HistoryManager(private val migrationHistoryRepository: MigrationHistoryRepository,
                          private val migrationRepository: MigrationRepository) {
    /**
     * Create a new history for migration
     * @param migration
     * @param step
     * @param data
     * @return
     */
    open fun startStep(migration: Migration?, step: StepEnum?, data: String?): MigrationHistory {
        val history = MigrationHistory()
            .step(step)
            .migration(migration)
            .date(Instant.now())
            .status(StatusEnum.RUNNING)
            .startTime(Instant.now())
        if (data != null) {
            history.data(data)
        }
        LOG.info("Start step $step")
        return migrationHistoryRepository.save(history)
    }

    /**
     * Update history
     * @param history
     */
    open fun endStep(history: MigrationHistory?, status: StatusEnum?, data: String? = null) {
        history?.status = status
        if (data != null) history?.data = data
        if (history?.startTime == null) {
            history?.executionTime = "N/A"
        } else {
            // Compute executionTime
            val execution = Instant.now().toEpochMilli() - history.startTime.toEpochMilli()
            history.executionTime = DateFormatter.toNiceFormat(execution)
        }
        migrationHistoryRepository.save(history!!)
        LOG.info("Finish step ${history.step} with status $status in ${history.executionTime}")
    }

    /**
     * Load eagerly a migration
     * @param migId Migration ID
     * @return migration loaded
     */
    @Transactional
    open fun loadMigration(migId: Long): Migration {
        val migration = migrationRepository.getById(migId)
        Hibernate.initialize(migration)
        Hibernate.initialize(migration.histories)
        Hibernate.initialize(migration.mappings)
        return migration
    }

    /**
     * Force flush for summary
     */
    open fun forceFlush() {
        migrationHistoryRepository.flush()
        migrationRepository.flush()
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(HistoryManager::class.java)
    }
}
